README
Dancer::Session::Cookie - encrypted cookie-based
session backend for Dancer framework.

This module implements a session engine for sessions stored entirely
in cookies. Usually only session id is stored in cookies and the
session data itself is saved in a database. This module allows to
avoid using external storage at all.

Since we cannot trust any data provided by client in cookies, we use
cryptography to ensure integrity.


Changes
Revision history for Dancer-Session-Cookie

    * Fix session destruction bug to ensure session is overwritten in
      the browser, not just no longer sent (David Golden)

0.15	2011-05-21

    * skip test 02-configfile.t if YAML is not present
      FIX smoker test failure.
      (Alexis Sukrieh)

0.14	2011-03-12
    
	[ Michael G. Schwern ]

    * Fix for Dancer > v1.3012 
    * Add support for session_secure to serve https only cookies. 
    * Add missing MYMETA.yml 
    * Make Dancer::Session::Cookie honor the session_name setting 
      added to Dancer::Session::Abstract
    * Add session_cookie_path to control the path of the cookie.

0.13    Sun Nov 28 17:19:58 MSK 2010
		Some documentation fixes. No functional change.

0.12    Thu Sep 16 13:29:09 MSD 2010
        Fix a warning when testing against newer Dancer. No functional
        change.

0.11    Wed Feb 17 16:12:21 MSK 2010
        Chase the ever-changing Dancer core :) No functional change.
        More tests.

0.1     Tue Feb  2 17:20:11 MSK 2010
        First version separated from the Dancer core. No functional
        change.


Makefile.PL
use strict;
use warnings;
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME                => 'Dancer::Session::Cookie',
    AUTHOR              => q{Alex Kapranoff <kappa@cpan.org>},
    VERSION_FROM        => 'lib/Dancer/Session/Cookie.pm',
    ABSTRACT_FROM       => 'lib/Dancer/Session/Cookie.pm',
    ($ExtUtils::MakeMaker::VERSION >= 6.3002
      ? ('LICENSE'=> 'perl')
      : ()),
    PL_FILES            => {},
    BUILD_REQUIRES      => {
        'Test::More'      => 0,
        'Test::NoWarnings'=> 0,
        'Test::Exception' => 0,
    },
    PREREQ_PM           => {
        'Crypt::CBC'      => 0,
        'String::CRC32'   => 0,
        'Crypt::Rijndael' => 0,
        'Dancer'          => 1.130,
    },
    dist                => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean               => { FILES => 'Dancer-Session-Cookie-* t/logs' },
);


lib/Dancer/Session/Cookie.pm
package Dancer::Session::Cookie;

use strict;
use warnings;
use base 'Dancer::Session::Abstract';

use Crypt::CBC;
use String::CRC32;
use Crypt::Rijndael;

use Dancer ();
use Dancer::Config 'setting';
use Storable     ();
use MIME::Base64 ();

use vars '$VERSION';
$VERSION = '0.15';

# crydec
my $CIPHER = undef;

sub init {
    my ($class) = @_;

    my $key = setting("session_cookie_key")  # XXX default to smth with warning
      or die "The setting session_cookie_key must be defined";

    $CIPHER = Crypt::CBC->new(
        -key    => $key,
        -cipher => 'Rijndael',
    );
}

sub new {
    my $self = Dancer::Object::new(@_);

    # id is not needed here because the whole serialized session is
    # the "id"
    return $self;
}

sub retrieve {
    my ($class, $id) = @_;

    my $ses = eval {
        # 1. decrypt and deserialize $id
        my $plain_text = _decrypt($id);

        # 2. deserialize
        $plain_text && Storable::thaw($plain_text);
    };

    $ses and $ses->{id} = $id;

    return $ses;
}

sub create {
    my $class = shift;
    return Dancer::Session::Cookie->new(id => 'empty');
}


# session_name was introduced to Dancer::Session::Abstract in 1.176
# we have 1.130 as the minimum
sub session_name {
    my $self = shift;
    return eval { $self->SUPER::session_name } || setting("session_name") || "dancer.session";
}

sub flush {
    my $self = shift;

    # 1. serialize and encrypt session
    delete $self->{id};
    my $cipher_text = _encrypt(Storable::freeze($self));

    # 2. write it into the cookie session ID
    $self->write_session_id( $self->{id} = $cipher_text );

    return 1;
}

sub destroy {
    my $self = shift;

    # 1. clear session data
    %$self = ();

    # 2. set dummy session ID and write it to the cookie to clear browser-side
    $self->{id} = 'empty';
    $self->write_session_id;

    return 1;
}

# Copied from Dancer::Session::Abstract and modified to add support
# for session_cookie_path
sub write_session_id {
    my ($class, $id) = @_;

    my $name = session_name();
    my %cookie = (
        name   => $name,
        value  => $id,
        path   => setting('session_cookie_path') || '/',
        domain => setting('session_domain'),
        secure => setting('session_secure'),
        http_only => defined(setting("session_is_http_only")) ?
                     setting("session_is_http_only") : 1,
    );
    if (my $expires = setting('session_expires')) {
        # It's # of seconds from the current time
        # Otherwise just feed it through.
        $expires = Dancer::Cookie::_epoch_to_gmtstring(time + $expires) if $expires =~ /^\d+$/;
        $cookie{expires} = $expires;
    }

    my $c = Dancer::Cookie->new(%cookie);
    Dancer::Cookies->set_cookie_object($name => $c);
}

sub _encrypt {
    my $plain_text = shift;

    my $crc32 = String::CRC32::crc32($plain_text);

    # XXX should gzip data if it grows too big. CRC32 won't be needed
    # then.
    my $res =
      MIME::Base64::encode($CIPHER->encrypt(pack('La*', $crc32, $plain_text)),
        q{});
    $res =~ tr{=+/}{_*-};    # cookie-safe Base64

    return $res;
}

sub _decrypt {
    my $cookie = shift;

    $cookie =~ tr{_*-}{=+/};

    $SIG{__WARN__} = sub {};
    my ($crc32, $plain_text) = unpack "La*",
      $CIPHER->decrypt(MIME::Base64::decode($cookie));
    return $crc32 == String::CRC32::crc32($plain_text) ? $plain_text : undef;
}

1;
__END__

=pod

=head1 NAME

Dancer::Session::Cookie - Encrypted cookie-based session backend for Dancer

=head1 SYNOPSIS

Your F<config.yml>:

    session: "cookie"
    session_cookie_key: "this random key IS NOT very random"

=head1 DESCRIPTION

This module implements a session engine for sessions stored entirely
in cookies. Usually only B<session id> is stored in cookies and
the session data itself is saved in some external storage, e.g.
database. This module allows to avoid using external storage at
all.

Since server cannot trust any data returned by client in cookies, this
module uses cryptography to ensure integrity and also secrecy. The
data your application stores in sessions is completely protected from
both tampering and analysis on the client-side.

=head1 CONFIGURATION

The setting B<session> should be set to C<cookie> in order to use this session
engine in a Dancer application. See L<Dancer::Config>.

A mandatory setting is needed as well: B<session_cookie_key>, which should
contain a random string of at least 16 characters (shorter keys are
not cryptographically strong using AES in CBC mode).

Here is an example configuration to use in your F<config.yml>:

    session: "cookie"
    session_cookie_key: "kjsdf07234hjf0sdkflj12*&(@*jk"

Compromising B<session_cookie_key> will disclose session data to
clients and proxies or eavesdroppers and will also allow tampering,
for example session theft. So, your F<config.yml> should be kept at
least as secure as your database passwords or even more.

Also, changing B<session_cookie_key> will have an effect of immediate
invalidation of all sessions issued with the old value of key.

B<session_cookie_path> can be used to control the path of the session
cookie.  The default is /.

The global B<session_secure> setting is honoured and a secure (https
only) cookie will be used if set.

=head1 DEPENDENCY

This module depends on L<Crypt::CBC>, L<Crypt::Rijndael>,
L<String::CRC32>, L<Storable> and L<MIME::Base64>.

=head1 AUTHOR

This module has been written by Alex Kapranoff.

=head1 SEE ALSO

See L<Dancer::Session> for details about session usage in route handlers.

See L<Plack::Middleware::Session::Cookie>,
L<Catalyst::Plugin::CookiedSession>, L<Mojolicious::Controller/session> for alternative implementation of this mechanism.

=head1 COPYRIGHT

This module is copyright (c) 2009-2011 Alex Kapranoff <kappa@cpan.org>.

=head1 LICENSE

This module is free software and is released under the same terms as Perl
itself.

=cut
